pep-0635.rst

```diff
@@ -0,0 +1,588 @@
+
+But this is not sufficient for all situations.  For example, a code
+generator may consume an AST, and have many operations where the
+generated code needs to vary based not just on the class of a node,
+but also on the value of some class attributes, like the ``BinOp``
+example above.  The Visitor pattern is insufficiently flexible for
+this: it can only select based on the class.
+
+For a complete example, see
+https://github.com/gvanrossum/patma/blob/master/examples/expr.py#L231
+
+TODO: Could we say more here?
+
+Pattern and functional style
+----------------------------
+
+Most Python applications and libraries are not written in a consistent
+OO style -- unlike Java, Python encourages defining functions at the
+top-level of a module, and for simple data structures, tuples (or
+named tuples or lists) and dictionaries are often used exclusively or
+mixed with classes or data classes.
+
+Pattern matching is particularly suitable for picking apart such data
+structures.  As an extreme example, it's easy to write code that picks
+a JSON data structure using ``match``.
+
+TODO: Example code.
+
+
+
+Motivation (Tobias' version)
+============================
+
+**This section should explain why we think pattern matching is a good
+addition for Python.**
+
+Since Python is a dynamically typed language, Python code frequently
+has to deal with data that comes in different forms and shapes.  This,
+in turn, gives rise to a high degree of versatility by favouring a
+'duck typing' style, imposing only the bare minimum of requirements
+on the form and shape of data, i.e. the structure of data objects.
+Nonetheless, we find that Python code is often sparkled with
+conditions depending on ``isinstance``, ``len``, ``getattr``,
+``hasattr``, etc.  Despite the benefits of 'duck typing', actual
+code still requires to query the format or type of an object in order
+to choose an appropriate processing action or to extract relevant bits
+of information.
+
+Unfortunately, the conditions for slightly more complex structures
+quickly grow into sequences of partly interdependent bits of
+structural tests, hurting readbility and maintability.  Pattern
+matching offers here a more direct scheme of expressing the minimal
+structure that an objects needs to have in order to allow for further
+processing.  Rather than writing a series of manual tests, patterns
+follow a *declarative* style, which is well-known for improved
+readability, maintability, and for delegating the burden of efficient
+execution on the compiler and interpreter.
+
+The concept of pattern matching is similar to regular expressions,
+where succinct patterns describe a textual structure.  A dedicated
+compiler then transforms these declarative patterns into highly
+efficient finite state machines.  In contrast to regular expressions,
+pattern matching targets Python objects rather than textual data,
+and builds on *decision trees* as the driving motor for finding a
```

Comment:
The reason for bringing in this idea of the decision tree is to offer a true alternative to `if`-`elif`-`else` chains and, moreover, to contrast it with lookup tables à la `switch`.  I really tried hard to not miss an opportunity to hint at the difference between pattern matching and `switch` ;-).

Reply:
Hmm, yes, you are quite right.  I probably wanted too much with all the different possible aspects.

I did not know that some regex implementations used bytecode rather than a FSMs.  This is probably an example of professional blindness as we often teach FSMs and then explain that regex are an example of them.

It's probably best to just ditch most of this and really concentrate more on the core aspects.